File: RadixSort.py
def counting_sort(arr, exp):
    n = len(arr)
    output = [0] * n  
    count = [0] * 10  

    
    for i in range(n):
        index = (arr[i] // exp) % 10
        count[index] += 1

    
    for i in range(1, 10):
        count[i] += count[i - 1]

    i = n - 1
    while i >= 0:
        index = (arr[i] // exp) % 10
        output[count[index] - 1] = arr[i]
        count[index] -= 1
        i -= 1

  
    for i in range(n):
        arr[i] = output[i]


def radix_sort(arr):
    
    max_val = max(arr)

   
    exp = 1  
    while max_val // exp > 0:
        counting_sort(arr, exp)
        exp *= 10



if __name__ == "__main__":
    
    try:
        array = list(map(int, input("Introdu numerele separate prin spațiu: ").split()))
        
        radix_sort(array)
        print("Lista sortata:")
        print(array)
    except ValueError:
        print("introdu doar numere întregi separate prin spațiu!")

File: Ciuraru_Alexandru/Radix_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void countSort(vector<int>& arr, int exp) {
    int n = arr.size();
    vector<int> output(n);
    vector<int> count(10, 0);

    for (int i = 0; i < n; i++) {
        count[(arr[i] / exp) % 10]++;
    }

    for (int i = 1; i < 10; i++) {
        count[i] += count[i - 1];
    }

    for (int i = n - 1; i >= 0; i--) {
        output[count[(arr[i] / exp) % 10] - 1] = arr[i];
        count[(arr[i] / exp) % 10]--;
    }

    for (int i = 0; i < n; i++) {
        arr[i] = output[i];
    }
}

void radixSort(vector<int>& arr) {
    int maxVal = 0;
    int n = arr.size();
    for (int i = 0; i < n; i++) {
        if (arr[i] > maxVal)
            maxVal = arr[i];
    }


    for (int exp = 1; maxVal / exp > 0; exp *= 10) {
        countSort(arr, exp);
    }
}

void printArray(const vector<int>& arr) {
    for (int num : arr) {
        cout << num << " ";
    }
    cout << endl;
}

int main() {
    vector<int> arr = { 170, 45, 75, 90, 802, 24, 2, 66 };


    radixSort(arr);

    printArray(arr);

    return 0;
}
